pkg/source/examples/simple_source/impl/simple_source.go

```diff
@@ -0,0 +1,121 @@
+package impl
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	sourcesdk "github.com/numaproj/numaflow-go/pkg/source"
+	"github.com/numaproj/numaflow-go/pkg/source/model"
+)
+
+// SimpleSource is a simple source implementation.
+type SimpleSource struct {
+	size     int64
+	messages []string
+	readIdx  int64
+	toAckSet map[int64]struct{}
+	lock     *sync.Mutex
+}
+
+// NewSimpleSource creates a SimpleSource by populating the messages array with random strings
+func NewSimpleSource(size int64) *SimpleSource {
+	messages := make([]string, size)
+	for i := int64(0); i < size; i++ {
+		messages[i] = strconv.FormatInt(i, 10)
+	}
+	return &SimpleSource{
+		size:     size,
+		messages: messages,
+		readIdx:  0,
+		toAckSet: make(map[int64]struct{}),
+		lock:     new(sync.Mutex),
+	}
+}
+
+// Pending returns the number of pending records.
+func (s *SimpleSource) Pending(_ context.Context) uint64 {
+	// The simple source always returns 0 to indicate no pending records.
+	return 0
+}
+
+// Read reads messages from the source and sends the messages to the message channel.
+// If the read request is timed out, the function returns without reading new data.
+// Right after reading a message, the function marks the offset as to be acked.
+func (s *SimpleSource) Read(_ context.Context, readRequest sourcesdk.ReadRequest, messageCh chan<- model.Message) {
+	// Handle the timeout specification in the read request.
+	ctx, cancel := context.WithTimeout(context.Background(), readRequest.TimeOut())
+	defer cancel()
+
+	// If we have un-acked data, or if we have read all the data,
+	// we return without reading any new data.
+	if len(s.toAckSet) > 0 || s.readIdx >= s.size {
+		return
```

Comment:
```suggestion
	// If we have un-acked data, we return without reading any new data.
	if len(s.toAckSet) > 0 {
		return
```

we support only unbounded stream so let's not give users ideas :-)